### ISSUE TITLE
Annotate constructor with annotation

### ISSUE THREAD

**Issue by VaclavPokorny** [immutables/immutables#456]
I was battling Spring MongoDB repositories to work with immutables and I got it working. For successful deserialization from Mongo the generated immutable type needs to have public constructor with `@PersistenceConstructor` annotation.

This works but looks little ugly:

```
@Immutable
@Style(of = "new", allParameters = true, jdkOnly = true, passAnnotations = {PersistenceConstructor.class})
abstract public class SampleInner {

    @PersistenceConstructor
    public SampleInner() {}

    abstract public String string();

    abstract public Integer integer();

    abstract public LocalDate localDate();

}
```

This would look much nicer:

```
@Immutable
@Style(of = "new", allParameters = true, jdkOnly = true, annotateConstructorWith = {PersistenceConstructor.class})
public interface SampleInner {

    String string();

    Integer integer();

    LocalDate localDate();

}
```

**Comment by elucash** [comment]
Agree about prettiness/ugliness, but that would be a very specific case as annotations can carry attributes, etc.

I've tried to hack spring data a bit to resolve public constructors: https://gist.github.com/elucash/a701a64ce608de6c6780052ab3b1eb00
But it's no so easy to plug that overriden mapper into the configuration.
It's should be also possible write completely standalone reader/writer converter,
but you may lose a lot of functionality/convenience in this way.

Given complications outlined above, I'm looking into a way to implement what you're proposing by trying to rationalize the ad-hoc nature of this.

(just btw, had you looking into our [MongoDB repository generation](http://immutables.github.io/mongo.html) ?
while limited to 2.x drivers version, it still kicks ass.
it's powered by http://github.com/michel-kraemer/bson4jackson and GSON generated type adapters, I've measured some day 2x-3x speed improvements in document mappings compared to Spring Data)

**Comment by VaclavPokorny** [comment]
I cannot use Immutables' MongoDB integration because driver version issues. Moreover, it would only solve this problem in case of this specific database.

Today I've also run into similiar problem with fields. For example the `org.springframework.data.mongodb.core.index.Indexed` annotation requires to be set directly on the field and not getter so there is no way of setting it on generated type.

I think there are 3 possible solutions to this:

1) Add springData switch to `@Style`. Code would look like this:
```
@Immutable
@Style(springData = true)
public interface SampleInner {
    @SpringData.Id
    String string();
    @SpringData.Indexed(unique=true)
    Integer integer();
    LocalDate localDate();
}
```
The style switch would do all the magic. It would be an ad-hoc solution that would cover only some use cases. Maintening it would require tracking future changes in Spring Data API. I would not recommend this.



2) Annotating annotations. The code would look like this:
```
@Immutable
@AnnotateConstructorWith(PeristenceConstructor.class)
@AnnotateConstructorParameterWith(name = "parameterWeWouldLikeToAnnotate", annotation = ConstructorParameterAnnotation.class)
public interface SampleInner {
    @AnnotateFieldWith(Id.class)
    String string();
    @AnnotateFieldWith(
            annotation = Indexed.class,
            parameters = {
                    @BooleanParameter(name = "unique", value = true)
                }
        )
    Integer integer();
    LocalDate localDate();
}
```
I think this would cover all imaginable cases but, of course, result will be a bit wordy.

3) Defer to an anotation mapping type. The code would look like this:
```
@Immutable
@AnnotationMapper(MyAnnotationMapper.class)
public interface SampleInner {
    String string();
    Integer integer();
    LocalDate localDate();
}

class MyAnnotationMapper implements ImmutablesAnnotationMapper {
    @Override
    public List<Annotation> processField(Field field) {...}
    @Override
    public List<Annotation> processConstructor(Method method) {...}
    
    ...
}
```
This could be the most flexible way. But I don't know if it could be done.

**Comment by VaclavPokorny** [comment]
Unfortunately, I found yet another issue. The Immutables' handling of Java 8 Optional is incompatible with Spring Data. The problem is that the getter and contructor are using `Optional<Type>` form but the field of generated class is just `Type`. Spring Data then throws an exception when instantiating the object because it does expect the parameter to be of `Type` and not of `Optional<Type>`. Can this behavior turned off somehow? Also, let's just rename this issue to something like "Spring Data compatibility".

**Comment by elucash** [comment]
a note about: `java.util.Optional<Type>` is not serializable and also because of the "advertised" way how people should use it (http://stackoverflow.com/questions/24547673/why-java-util-optional-is-not-serializable-how-to-serialize-the-object-with-suc) people were strongly suggesting to implement it using a nullable field. This is just as an example of how a field might be not of the same size and shape as the declared attribute type. You can use Guava or Fugue optional to have it stored the same as the field.

I do not believe that neurosurgery on the fields - injecting annotations - will get us anywhere. I'll try to see if there's a way to employ Spring Data abstractions to encode immutable objects: `MongoPersistentEntity` etc.

**Comment by elucash** [comment]
Since this ticket was reported, we have more or less working annotation injection facility (somewhat explained in #475) to be able to "push" annotations down to generated elements. Closing this, and any tickets wrt annotation injections are welcome.
